Feat/ks add pivot tables to react tables rebased

kristiansigston opened the pull request:
Add pivot tables to react

Devmode default false

It needs testing against known values
When no aggregate data field is selected aggregate function field should show all. Selection here should narrow aggregate data field selection to appropriate columns
Currently only works the other way.

Selecting an aggregate data field will enable extra count choices.

More refactoring required. More tests required.

I need to work with one of you to adjust this as I'm not sure what else I might break if I change how this is setup.

kristiansigston committed:
add pivot table functionality

extract pivot fields component

kristiansigston committed:
fix table merge issues

tgsoverly committed:
Add csv rows

kristiansigston committed:
add parent component for aggregatable fields

kristiansigston committed:
default devtools to false

tgsoverly commented on viewer/src/components/items/table-components/DuckDBStatus.jsx:
```diff
@@ -0,0 +1,20 @@
+      {duckDBStatus.state === "loading" && (
+        <>
+          <div className="animate-spin h-4 w-4 border-2 border-red-200 rounded-full border-t-transparent" />
+          <span className="text-sm text-red-700">Loading DuckDB...</span>
```
When does this get shown?

tgsoverly commented on viewer/src/components/items/table-components/PivotFields.jsx:
```diff
@@ -0,0 +1,75 @@
+        const accessorKey = rawAccessorKey.replace(/[.]/g, "_"); // Sanitize accessor key for react rows
+        const sampleRows = tableData.slice(0, 100)
+        
+        const isAggregatable = detectColumnType(sampleRows, accessorKey) === NUMBER;
```
Only numbers are aggregatable?  Not dates?

tgsoverly commented on viewer/src/components/items/Table.jsx:
```diff
@@ -87,40 +121,69 @@ const Table = ({ table, project, itemWidth, height, width }) => {
+          },
+        })),
+
+    data: isPivoted ? pivotedData : tableData,
```
I have seen this logic a couple times, worth pulling into a new variable?

kristiansigston replied:
I am not seeing it repeated. It's doing different things based on the pivoted being active.
What are you thinking?

tgsoverly commented on viewer/src/components/items/table-helpers/sanitizeColumnName.js:
```diff
@@ -0,0 +1,11 @@
+
+  const strName = typeof name !== "string" ? String(name) : name;
+
+  return strName.replace(/[^a-zA-Z0-9_]/g, "_").replace(/\s+/g, "_");
```
easily tested so we know what it does.

tgsoverly commented on viewer/src/components/items/table-helpers/detect-column-type/detectColumnType.js:
nit: I don't think these extra folders that have the - in them help clean it up. "table-helpers" is probably enough.

kristiansigston replied:
ok

kristiansigston replied:
The folders were so that the test and file were together

kristiansigston replied:
Would you still like me to remove them?

tgsoverly commented on viewer/src/components/items/table-components/pivot-column-selection/PivotColumnSelection.jsx:
```diff
@@ -0,0 +1,214 @@
+  useEffect(() => {
+    if (db && tableData && tableData.length > 0 && !isLoadingDuckDB && !dataLoadedRef.current) {
+      // Mark as loaded to prevent reload
+      console.log("Loading data to DuckDB...", tableData.length, "rows");
```
Lets remove the console.logs

kristiansigston replied:
They were for info but I can remove them
They should have said "info" 
Consider them gone.

kristiansigston replied:
Eww. Some of those were definitely not supposed to be left in

tgsoverly commented on viewer/src/components/items/Table.test.jsx:
```diff
@@ -157,12 +178,16 @@ test('handles number values in cells', async () => {
+    expect(screen.getByText("123")).toBeInTheDocument();
   });
   expect(screen.getByText('12,345,678,901,234,568')).toBeInTheDocument();
   expect(screen.getByText('1,234,567,890.123')).toBeInTheDocument();
```
I would love a test flexing the pivot functionality if possible.

kristiansigston replied:
Yes. That requires using duckdb-wasm sql inside the test env.
Setting that up would take some time and investigation.

Do we have full run integration tests?

kristiansigston commented on viewer/src/components/items/duckdb-wasm-init/duckDBWasmInit.spec.js:
I will rename

kristiansigston replied:
Done

kristiansigston commented on viewer/src/components/items/table-helpers/create-sanitized-value-sql/createSanitizedValueSql.test.js:
```diff
@@ -0,0 +1,42 @@
+describe("createSanitizedValueSql", () => {
+  it("should create a SQL expression that sanitizes a value field for numeric operations", () => {
+    const fieldName = "value_field";
+    const expectedSql = `
```
This makes sure no one accidentally changes the SQL file.

kristiansigston commented on viewer/src/hooks/__tests__/batchInsertData.test.js:
ok. My bad.

kristiansigston replied:
Perhaps I was confused with core/django

kristiansigston commented on viewer/src/setupTests.js:
```diff
@@ -3,8 +3,60 @@
 // expect(element).toHaveTextContent(/react/i)
```
What is this comment? Remove

kristiansigston commented on viewer/src/components/items/duckdb-wasm-init/DuckDBCacheStatus.jsx:
```diff
@@ -0,0 +1,113 @@
+
+export default DuckDBCacheStatus;
+
+// not currently using this
```
There is a prettier setting that auto formats an extra line at the end of a file to remove that sign.
Would you like me to add it.

kristiansigston commented on visivo/viewers/local/index.html:
```diff
@@ -28,6 +28,10 @@
     <link rel="stylesheet" crossorigin href="/assets/index-fHoPSZ5b.css">
   </head>
+  <title>Visivo</title>
+  <script type="module" crossorigin src="/assets/index-DReODLCq.js"></script>
```
Hmm. That might have been a merge fail. on my part.

kristiansigston replied:
Fixed

kristiansigston committed:
pr update requests

remove errant console.logs
make test naming and placement consistent
remove unnecesary file folder nesting

kristiansigston committed:
remove packageManager

add debugger to vs code for visivo commands

kristiansigston committed:
update tests to reflect removed console info,log,errors

kristiansigston committed:
remove unused variables

and apply correct linting

kristiansigston commented:
@tgsoverly 
Hi Tim.

I have merged with Main
I can still pivot fine.

What yarn version should we be using?

We might need to pair on this a bit

I need to understand you dashboard setup as it seems you are using calculated set heights with grid.

What browser are you using?
What node version are you using?

![image](https://github.com/user-attachments/assets/de5947d6-1f6b-4dcb-94ee-9fc8b305d5bb)

kristiansigston committed:
remove dead space from dropdown

persist all but agregatable field selection

kristiansigston committed:
update dependency array for array field persistence

run yarn install

kristiansigston commented on viewer/package.json:
```diff
@@ -4,6 +4,7 @@
   "private": true,
   "proxy": "http://0.0.0.0:8000",
   "dependencies": {
+    "@duckdb/duckdb-wasm": "^1.29.1-dev132.0",
```
Do we need this if we don't use it?
Did I add this?

kristiansigston committed:
initialise pivot options on first mount

but don't unmount while page active

kristiansigston committed:
remove unused imports
